File: src/templates/template.c
<%!
import sympy
from sympy.printing import ccode

%>

/* Parameters */
% for key in model.parameters:
static const double ${key + ' = ' + ccode(model.parameters[key]) + ';'}
% endfor

/* Algebraic equations */
void algebraic(double *alg, double *u, double t)
{
% for key in model.ode:
    ${key + ' = ' + 'u[' + model.idx_state[key].__str__() + '];'}
% endfor

% for key in model.algebraic:
    ${key + ' = ' + ccode(model.algebraic[key]) + ';'}
% endfor 

% for key in model.algebraic:
    ${'alg[' + model.idx_algebraic[key].__str__() + '] = ' + key + ';' }
% endfor
}

/* State Variable ODEs */
void state(double *du, double *u, double t, double *alg)
{
    /* Extract algebraic and state variables */
% for key in model.algebraic:
    ${key + ' = ' + 'alg[' + model.idx_algebraic[key].__str__() + '];'}
% endfor
% for key in model.ode:
    ${key + ' = ' + 'u[' + model.idx_state[key].__str__() + '];'}
% endfor
    
    /* ODE Right hand sides */
% for key in model.ode:
    ${'du[' + model.idx_state[key].__str__() + '] = ' + ccode(
            model.ode[key]) + ';'}
% endfor
}

File: src/templates/template.m
<%!
import sympy
from sympy.printing import octave_code
%>

function params = parse_inputs(varargin)
p = inputParser()
% for key in model.parameters:
p.addParameter('${key}', ${octave_code(model.parameters[key])});
% endfor
p.parse(varargin{:})
params = p.Results;
end


% for key in model.parameters:
${key + ' = ' + octave_code(model.parameters[key]) + ';'}
% endfor

function alg = algebraic(u, t)
% for key in model.ode:
    ${key + ' = ' + 'u(' + (model.idx_state[key] + 1).__str__() + ', :);'}
% endfor

% for key in model.algebraic:
    ${key + ' = ' + octave_code(model.algebraic[key]) + ';'}
% endfor 

% for key in model.algebraic:
    alg = zeros(${}, size(u, 2));
    ${'alg(' + (model.idx_algebraic[key] + 1).__str__() + ', :) = ' + key + ';' }
% endfor
end

function du = ode(u, t, alg)
%% Extract algebraic and state variables
% for key in model.algebraic:
    ${key + ' = ' + 'alg(' + (model.idx_algebraic[key] + 1).__str__() + ', :);'}
% endfor
% for key in model.ode:
    ${key + ' = ' + 'u(' + (model.idx_state[key] + 1).__str__() + ', :);'}
% endfor
    
%% ODE Right hand sides 
    du = zeros(size(u));
% for key in model.ode:
    ${'du(' + (model.idx_state[key] + 1).__str__() + ', :) = ' + octave_code(
            model.ode[key]) + ';'}
% endfor
end
